### DISCUSSION TITLE
failed to register IUnitOfWork using containerRegistry

### DISCUSSION THREAD

**Discussion by ErwinDraconis** [PrismLibrary/Prism#2330] [WPF] [answered]
Hello,
after successfully registering IUnitOfWork using 'Microsoft.Extensions.DependencyInjection 5.0.0' :

![image](https://user-images.githubusercontent.com/42243889/106755238-2b8b2480-662e-11eb-886d-74e3934aa01c.png)

![image](https://user-images.githubusercontent.com/42243889/106755300-43fb3f00-662e-11eb-8b21-2bf74ff7b892.png)

 i tried using prism to register the same interface this time with containerRegistry.unfortunatly there is no AddDatabase method, so i tried the following 

![image](https://user-images.githubusercontent.com/42243889/106755492-7d33af00-662e-11eb-8aad-469a8c9e6f56.png)

but i get an error  : ResolutionFailedException: Resolution failed with error: No public constructor is available for type Microsoft.EntityFrameworkCore.DbContextOptions.

thank you for the help !

**Comment by dansiegel** [comment, marked as answer]
@ErwinDraconis while Prism's container can generally resolve concrete types there are a few things that you probably want to adjust here...

First of all, unless you're a DI guru, you really want to avoid using Scoped Services. That is a VERY advanced topic. Remember that Microsoft.Extensions.DependencyInjection was purpose built for ASP.NET where Scoped lifetimes make a lot of sense as each request represents a different user. In the Desktop or Mobile world, you have a single user, and more than likely want a Singleton or a Transient. For your IUnitOfWork and IUserRepository you probably want RegisterSingleton not RegisterScoped.

By not registering the concrete DatabaseContext you are implicitly registering it as a Transient Service, or in other words each time it is resolved you will get a brand new instance which may or may not be good for your purposes. However you have a second and more important issue which is what is leading to the error that you see here. By providing no registrations there you haven't told it how to construct the DbContextOptions which you undoubtedly have as a constructor argument in your DatabaseContext. Essentially you need to create a DbContextOptionsBuilder where you would use the UseSqlite extension and then register the options instance from the builder.

**Reply by ErwinDraconis** [reply]
@dansiegel Thank you so much for your help. based on your suggestion it worked, hier is the implementation i used :
var optionsBuilder = new DbContextOptionsBuilder<DatabaseContext>(); var options = optionsBuilder .UseSqlite(@"Data Source=DataBase\\Runnin.db") .Options; containerRegistry.RegisterSingleton<DatabaseContext>(() => new DatabaseContext(options)); containerRegistry.RegisterSingleton<IUnitOfWork, UnitOfWork>(); containerRegistry.RegisterSingleton<IUserRepository, UserRepository>();

**Reply by ErwinDraconis** [reply]
Hello @dansiegel may be this should go in other thread,but since its talking about the same topic i decide to put it hier 👍
now i want to register a logger in the container, i use log4net, its working but the problem is that the log context is always 'App' :
![image](https://user-images.githubusercontent.com/42243889/107003538-43c58580-678d-11eb-9574-386d1bad1cf0.png)

injection in the view model

![image](https://user-images.githubusercontent.com/42243889/107003575-50e27480-678d-11eb-9c12-5f3089fcf1c7.png)

the output in text file is always pointing to App instead of the curent context which is in this case the viewModel

![image](https://user-images.githubusercontent.com/42243889/107003625-6061bd80-678d-11eb-8ffa-62fb1f3344fd.png)

**Reply by dansiegel** [reply]
That's probably best as it's own discussion

**Reply by ErwinDraconis** [reply]
yes make sense !
